### PR TITLE
Things that came up in testing group while testing quotes

### LINKED ISSUE #1654: move directly to first-unread message
with https://github.com/deltachat/deltachat-android/pull/1641, a chat is opened at the position of the first unread message.

however, there is a short flickering as the chat is opened at the end first and scrolls up afterwards. would be nice if we can get rid of that.

### PR DESCRIPTION
**Pull request by Hocuri**
See individual commits
fix https://github.com/deltachat/deltachat-android/issues/1653
fix #1654

### CODE REVIEW

File: src/com/b44t/messenger/DcMsg.java [hunk 1]
```diff
@@ -71,6 +73,21 @@ public boolean equals(Object other) {
         return this.getId()==that.getId() && this.getId()!=0;
     }
 
+    /**
+     * If given a message, calculates the position of the message in the chat
+     */
+    public static int getMessagePosition(DcMsg msg, ApplicationDcContext dcContext) {
+        int msgs[] = dcContext.getChatMsgs(msg.getChatId(), 0, 0);
+        int startingPosition = -1;
+        for(int i=0; i< msgs.length; i++ ) {
```
Comment (r10s) [inline src/com/b44t/messenger/DcMsg.java:82]:
```suggestion
        for (int i = 0; i < msgs.length; i++) {
```

File: src/com/b44t/messenger/DcMsg.java [hunk 2]
```diff
@@ -71,6 +73,21 @@ public boolean equals(Object other) {
...
+            if(msgs[i] == msg.getId()) {
```
Comment (r10s) [inline src/com/b44t/messenger/DcMsg.java:83]: i would tend to get `msg.getId()` out of the loop and not only rely on the optimizer. a chat with some 1000 messages may result in lots of ffi calls, done before the chat is displayed, on slow devices, this may be some 100 ms.

File: src/org/thoughtcrime/securesms/MyLayoutManager.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+import androidx.recyclerview.widget.LinearLayoutManager;
+import androidx.recyclerview.widget.RecyclerView;
+
+import org.thoughtcrime.securesms.util.ViewUtil;
+
+/**
+ * Like LinearLayoutManager but you can set a starting position
+ */
+class MyLayoutManager extends LinearLayoutManager {
```
Comment (r10s) [inline src/org/thoughtcrime/securesms/MyLayoutManager.java:14]: can we have a more describing name for that?

also, overwriting a central function - why is that needed? this may cause troubles on different implementations.

i just had a look - signal uses SmoothScrollingLinearLayoutManager - is that comparable?

Comment (Hocuri) [reply src/org/thoughtcrime/securesms/MyLayoutManager.java:14]: No, I guess that Signal uses SmoothScrollingLinearLayoutManager in order to be able to set the speed of smooth scrolling. We need a custom LayoutManager in order to be able to set the starting position ( 5c5e267 ). I don't know how Signal does this but their code has diverged a lot from ours in this area, so that I decided it's easier to implement this myself (would have been at least some 100 lines to pull in)

I thought about renaming, too, but didn't know a good name. SetStartingPositionLinearLayoutManager?

Comment (r10s) [reply src/org/thoughtcrime/securesms/MyLayoutManager.java:14]: thanks for explaining. and yes, `SetStartingPositionLayoutManager` looks good to me.  maybe a bit long, however, it is not used that often and make things much clearer.

**Review by r10s — APPROVED** [review]
i suggest to rename from MyLayoutManager to SetStartingPositionLinearLayoutManager, however, i already approve, so you can merge directly after rename (in case you agree :)

**Comment by r10s** [comment]
@Hocuri great, thanks for merging! quotes are so much better now!
